static/js/template/controllers/template.controller.js

```diff
@@ -1,349 +1,361 @@
-        self.items.push(item);
+        function onOver(event, ui) {
+        }
 
-      } else if (item_type === 'select') {
+        function generateId() {
+            return 'id' + ++idGenIndex;
+        }
 
-        var item = {
-          id_string: curId,
-          name: 'select_placeholder' + curId,
-          type: item_type,
-          width: 100,
-          height: 100,
-          values: null,
-          role: 'input',
-          sub_type: 'div',
-          layout: 'column',
-          data_source: null,
-          position: self.items.length + 1
-        };
-        self.items.push(item);
-      } else if (item_type === 'labeled_input') {
-        var item = {
-          id_string: curId,
-          name: 'labeled_input' + curId,
-          type: item_type,
-          width: 100,
-          height: 100,
-          values: 'Label1',
-          role: 'input',
-          sub_type: 'h4',
-          layout: 'column',
-          data_source: null,
-          position: self.items.length + 1
-        };
-        self.items.push(item);
-      }
-      sync();
-    }
+        function generateRandomTemplateName() {
+            var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+            var random = _.sample(possible, 8).join('');
+            return 'template_' + random;
+        }
 
-    function onOver(event, ui) {
-    }
+        function sync() {
+            $scope.project.currentProject.template = {
+                name: self.templateName,
+                items: self.items
+            }
+        }
 
-    function generateId() {
-      return 'id' + ++idGenIndex;
-    }
+        //Show Modal Pop-up of the Task Design Output
+        function showTaskDesign(previewButton) {
+            update_item_data();
+            $mdDialog.show({
+                template: '<md-dialog>' +
+                '<md-dialog-content>' +
+                '<h3><span ng-bind="project.currentProject.name"></span></h3>' +
+                '<span ng-bind="project.currentProject.description"></span>' +
+                '<md-divider></md-divider>' +
+                '<ul ng-model="template.items" class="no-decoration-list">' +
+                '<li class="template-item" ng-repeat="item in template.items_with_data">' +
+                '<div md-template-compiler="template.buildHtml(item)"></div>' +
+                '</li>' +
+                '</ul>' +
+                '</md-dialog-content>' +
+                '</md-dialog>',
+                parent: angular.element(document.body),
+                scope: $scope,
+                targetEvent: previewButton,
+                preserveScope: true,
+                clickOutsideToClose: true
+            });
+        }
 
-    function generateRandomTemplateName() {
-      var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
-      var random = _.sample(possible, 8).join('');
-      return 'template_' + random;
-    }
+        function update_item_data() {
```

Comment:
function which updates the data field, it's a deep copy because we dont want to change the original items